File: ts/day2.ts
const data: string = await Deno.readTextFile("../data/day2.txt")
const lines: Array<string> = data.split("\n")

interface PasswordEntry {
    min: number, 
    max: number, 
    letter: string,
    password: string
}

const lineToPE = function (line: string): PasswordEntry {
    const chunks = line.split(":")
    const minMaxLetter = chunks[0].split(" ")
    const minMax = minMaxLetter[0].split("-")
    const pe = {
        min: Number(minMax[0]), 
        max: Number(minMax[1]),
        letter: minMaxLetter[1],
        password: chunks[1].trim()
    }
    return pe
}

const checkValidPE = function (pe: PasswordEntry): boolean {
    let charCount = 0
    Array.from(pe.password).forEach(char => {
        if (char == pe.letter) charCount += 1
    })
    return (charCount >= pe.min && charCount <= pe.max) 
}

const entries = lines.map(lineToPE)
const validPasswords = entries.filter(pe => checkValidPE(pe) === true)

console.log("part1", validPasswords.length)


// Part 2


interface PasswordEntryT {
    p1: number, 
    p2: number, 
    letter: string,
    password: string
}

const lineToPET = function (line: string): PasswordEntryT {
    const chunks = line.split(":")
    const minMaxLetter = chunks[0].split(" ")
    const minMax = minMaxLetter[0].split("-")
    const pe = {
        p1: Number(minMax[0]) - 1, 
        p2: Number(minMax[1]) - 1,
        letter: minMaxLetter[1],
        password: chunks[1].trim()
    }
    return pe
}


const checkValidPET = function (pe: PasswordEntryT): boolean {
    return ((pe.password[pe.p1] === pe.letter && pe.password[pe.p2] != pe.letter) || (pe.password[pe.p1] != pe.letter && pe.password[pe.p2] === pe.letter)) 
}

const entriesT = lines.map(lineToPET)
const validPasswordsT = entriesT.filter(entry => checkValidPET(entry) === true)

console.log("part2", validPasswordsT.length)

File: python/day2.py
from collections import namedtuple


PasswordEntry = namedtuple("PasswordEntry", "min, max, letter, password")


def parse_line(line):
    chunks = line.split(":")
    password = chunks[1].strip()
    rule_chunk = chunks[0].split(" ")
    letter = rule_chunk[1]
    rules = rule_chunk[0].split("-")
    r1 = int(rules[0])
    r2 = int(rules[1])
    return PasswordEntry(r1, r2, letter, password)


def parse_passwords(fn):
    with open(fn, "r") as f:
        lines = f.readlines()
        pass_entries = [parse_line(line.strip()) for line in lines]
    return pass_entries


def check_valid_pass(pe):
    count = pe.password.count(pe.letter)
    return (count >= pe.min) & (count <= pe.max)


def part_one():
    pass_entries = parse_passwords("../data/day2.txt")
    valid_passwords = list(filter(check_valid_pass, pass_entries))
    print(len(valid_passwords))


def check_valid_passT(pe):
    return (
        (pe.password[pe.min - 1] == pe.letter) & (pe.password[pe.max - 1] != pe.letter)
    ) | (
        (pe.password[pe.min - 1] != pe.letter) & (pe.password[pe.max - 1] == pe.letter)
    )


def part_two():
    pass_entries = parse_passwords("../data/day2.txt")
    valid_passwords = list(filter(check_valid_passT, pass_entries))
    print(len(valid_passwords))


if __name__ == "__main__":
    # part_one()
    part_two()
